File: problems/rotate_array.py
"""189. Rotate Array"""

from typing import List


class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        length = len(nums)
        # Rotating k^n times is the same as rotating k times
        k %= length

        # Iterative
        # rotated = [0] * length
        # for i in range(length):
        #     rotated[(i + k) % length] = nums[i]
        # for i in range(length):
        #     nums[i] = rotated[i]

        # Rotating to the right is the same as
        # reversing both the elements before k and the elements after k
        # in distinct groups
        # if k != 0:
        #     # Reverse each group of elements
        #     nums[:k], nums[k:] = nums[-k:], nums[:-k]

        def reverse(i: int, x: int):
            while i < x:
                nums[i], nums[x] = nums[x], nums[i]
                i += 1
                x -= 1
         # Reverse the entire array
        reverse(0, length - 1)
        # Reverse the elements before k
        reverse(0, k - 1)
        # Reverse the remaining elements after k
        reverse(k, length - 1)

File: problems/java/rotate_array.java
/*
 * 189. Rotate Array
 */
class Solution {

  public void rotate(int[] nums, int k) {
    // BRUTE FORCE
    // int[] nums_right_rotate = new int[nums.length];
    // for (int i = 0; i < nums.length; i++) {
    //   nums_right_rotate[(i + k) % nums.length] = nums[i];
    // }
    // for (int i = 0; i < nums_right_rotate.length; i++) {
    //   nums[i] = nums_right_rotate[i];
    // }

    // TWO POINTER
    k %= nums.length;
    reverse(nums, 0, nums.length - 1);
    reverse(nums, 0, k - 1);
    reverse(nums, k, nums.length - 1);
  }

  public void reverse(int[] nums_rotate_right, int start, int end) {
    while (start < end) {
      int tmp = nums_rotate_right[start];
      nums_rotate_right[start] = nums_rotate_right[end];
      nums_rotate_right[end] = tmp;
      start++;
      end--;
    }
  }
}
